### ISSUE TITLE
Agents with websocket connection enabled often move to lost contact status

### ISSUE THREAD

**Issue by jyotisingh** [gocd/gocd#3438]
##### Issue Type
- Bug Report

##### Environment
Performance server

##### Description

Agents with websocket connection enabled often move to lost contact status.  No error reported in agent logs.

**Comment by jyotisingh** [comment]
@ibnc , @arvindsv - While running the performance pipeline with websocket connection enabled, noticed the following:

Many of the agents (running with websockets enabled) lost contact with the perf server. Spent a little time with the logs and code. I think I know what's happening. 
Many of the agents have this entry in their logs:
```
2017-04-28 06:32:16,360 [WebSocketClient@267572650-3315] DEBUG thoughtworks.go.agent.AgentWebSocketClientController:266 - [172.16.38.74/172.16.38.74:8154] closed. code: 1001, reason: Idle Timeout
```

So, the idle-timeout for websocket connections is set to 60secs. If the agent is sitting idle for about 60 secs, then the connection times out. When the connection drops, [this](https://github.com/gocd/gocd/blob/a83d1a5858bcd9f6a9d475940cddd38bba4073c1/server/src/com/thoughtworks/go/server/websocket/AgentRemoteHandler.java#L134) line in `AgentRemoteHandler` would mark the agent as 'lost contact'. The loop thread on the agent runs every 10secs, so about 10 secs later when the 'loop' thread is active again, it finds that the web-socket connection has dropped, so it attempts to re-connect ([code](https://github.com/gocd/gocd/blob/a83d1a5858bcd9f6a9d475940cddd38bba4073c1/agent/src/com/thoughtworks/go/agent/AgentWebSocketClientController.java#L105)). 

```
2017-04-28 06:32:24,463 [loopThread] INFO  thoughtworks.go.agent.WebSocketClientHandler:53 - Connecting to websocket endpoint: wss://172.16.38.74:8154/go/agent-websocket
```

If you go to the agents page on dashboard, and wait for a bit, you would see the agent come up over there. These are also able to pick and execute jobs. All agents go through this cycle. 

We always noticed that many agents had lost contact when websockets was enabled, but did not really notice that they came back too and something else lost contact. So, at this point it seems like things are working as expected, no real issue there as far as websocket connections are concerned. However, since many agents could remain idle on a given system, many of them would always be marked as 'lost contact' creating a false perception of an issue with the agents. For legacy communication, the server waited for 10mins before marking the agent as lost contact. Although with web-sockets we have the capability to know instanteneouly when the connection dropped, may be it would make sense to retain the original behavior. What do you guys think?

**Comment by jyotisingh** [comment]
There is another issue around job messages which require acknowledgement, one such example: 

Agent [sends](https://github.com/gocd/gocd/blob/a83d1a5858bcd9f6a9d475940cddd38bba4073c1/agent/src/com/thoughtworks/go/agent/BuildRemoteRepositoryAdapter.java#L41) this message to server, registers a [callback](https://github.com/gocd/gocd/blob/a83d1a5858bcd9f6a9d475940cddd38bba4073c1/agent/src/com/thoughtworks/go/agent/WebSocketSessionHandler.java#L74) which would be invoked when the server responds with an acknowledment. :
```
2017-04-28 07:19:35,137 [pool-1-thread-5] DEBUG thoughtworks.go.agent.WebSocketSessionHandler:68 - Session[172.16.38.74/172.16.38.74:8154] send message: Message{action=reportCompleting, data={"agentRuntimeInfo":{"identifier":{"hostName":"blrstdgoprf04.thoughtworks.com","ipAddress":"10.132.4.124","uuid":"7af45e15-398a-454e-a70f-d08041f1995b"},"runtimeStatus":"Cancelled","buildingInfo":{"buildingInfo":"","buildLocator":""},"location":"/home/go/agent/go-agents/agent-12","usableSpace":1460504633344,"operatingSystemName":"CentOS 6.6 Final","cookie":"76d9d40b-9437-4b9a-99e7-af129e6de321","supportsBuildCommandProtocol":false},"jobIdentifier":{"pipelineName":"gocd.perf35","pipelineCounter":15,"pipelineLabel":"15","stageName":"default","buildName":"defaultJob","buildId":11257,"stageCounter":"1"},"result":"Cancelled"}, acknowledgementId=89509d35-acf5-4181-beeb-24e0cb8e679e}
```

Server responds with an acknowledgment:

```
2017-04-28 07:19:35,622 [WebSocketClient@1401642976-3701] DEBUG thoughtworks.go.agent.AgentWebSocketClientController:249 - [172.16.38.74/172.16.38.74:8154] message: Message{action=acknowledge, data="89509d35-acf5-4181-beeb-24e0cb8e679e", acknowledgementId=f555775e-fc3c-4817-9f53-38419c07d42d}
```
A little later agent lost-contact with the server, and then reconnects. While reconnecting, it clears out all the registered `callbacks`. The server resends? the message. 

Agent barfs [here](https://github.com/gocd/gocd/blob/a83d1a5858bcd9f6a9d475940cddd38bba4073c1/agent/src/com/thoughtworks/go/agent/WebSocketSessionHandler.java#L108):

```
2017-04-28 07:26:44,422 [pool-1-thread-4] DEBUG thoughtworks.go.agent.WebSocketSessionHandler:107 - Acknowledging 89509d35-acf5-4181-beeb-24e0cb8e679e
2017-04-28 07:26:44,422 [pool-1-thread-4] ERROR thoughtworks.go.agent.AgentWebSocketClientController:258 - Unexpected error while processing message[Message{action=acknowledge, data="89509d35-acf5-4181-beeb-24e0cb8e679e", acknowledgementId=f555775e-fc3c-4817-9f53-38419c07d42d}]: null
java.lang.NullPointerException
	at com.thoughtworks.go.agent.WebSocketSessionHandler.acknowledge(WebSocketSessionHandler.java:108)
	at com.thoughtworks.go.agent.AgentWebSocketClientController.process(AgentWebSocketClientController.java:151)
	at com.thoughtworks.go.agent.AgentWebSocketClientController$1.run(AgentWebSocketClientController.java:254)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

Basically `callbacks` does not have an entry for the acknowledmentId `89509d35-acf5-4181-beeb-24e0cb8e679e` so `callbacks.remove(acknowledgementId)` return `null` and hence the `NullPointerException`.

Need to go through all such messages and see what could go wrong if there are any side-effects if the callback did not go through.

**Comment by jyotisingh** [comment]
This is this too, need to dig into it further:

```
2017-04-28 07:26:43,028 [pool-1-thread-3] DEBUG thoughtworks.go.work.DefaultGoPublisher:128 - [go] Job is cancelled gocd.perf492/11/default/1/defaultJob on blrstdgoprf04.thoughtworks.com [/home/go/agent/go-agents/agent-12]
2017-04-28 07:26:43,028 [pool-1-thread-3] INFO  thoughtworks.go.work.DefaultGoPublisher:94 - Agent [blrstdgoprf04.thoughtworks.com, 10.132.4.124, 7af45e15-398a-454e-a70f-d08041f1995b] is reporting status [Completed] to Go Server for Build [gocd.perf492/11/default/1/defaultJob/12320]
2017-04-28 07:26:43,028 [pool-1-thread-3] DEBUG thoughtworks.go.agent.WebSocketSessionHandler:68 - Session[172.16.38.74/172.16.38.74:8154] send message: Message{action=reportCurrentStatus, data={"agentRuntimeInfo":{"identifier":{"hostName":"blrstdgoprf04.thoughtworks.com","ipAddress":"10.132.4.124","uuid":"7af45e15-398a-454e-a70f-d08041f1995b"},"runtimeStatus":"Building","buildingInfo":{"buildingInfo":"gocd.perf525/18/default/1/defaultJob","buildLocator":"gocd.perf525/18/default/1/defaultJob"},"location":"/home/go/agent/go-agents/agent-12","usableSpace":1460329693184,"operatingSystemName":"CentOS 6.6 Final","cookie":"76d9d40b-9437-4b9a-99e7-af129e6de321","supportsBuildCommandProtocol":false},"jobIdentifier":{"pipelineName":"gocd.perf492","pipelineCounter":11,"pipelineLabel":"11","stageName":"default","buildName":"defaultJob","buildId":12320,"stageCounter":"1"},"jobState":"Completed"}, acknowledgementId=da32ff3d-5189-4416-b246-c920ba6f7f98}
2017-04-28 07:26:43,028 [pool-1-thread-3] ERROR go.remote.work.BuildWork:123 - New error occurred during error handling:
build will be rescheduled when agent starts asking for work again
org.eclipse.jetty.websocket.api.WebSocketException: RemoteEndpoint unavailable, outgoing connection not open
	at org.eclipse.jetty.websocket.common.WebSocketSession.getRemote(WebSocketSession.java:252)
	at com.thoughtworks.go.agent.WebSocketSessionHandler.send(WebSocketSessionHandler.java:69)
	at com.thoughtworks.go.agent.WebSocketSessionHandler.sendWithCallback(WebSocketSessionHandler.java:89)
	at com.thoughtworks.go.agent.WebSocketSessionHandler.sendAndWaitForAcknowledgement(WebSocketSessionHandler.java:74)
	at com.thoughtworks.go.agent.BuildRepositoryRemoteAdapter.reportCurrentStatus(BuildRemoteRepositoryAdapter.java:35)
	at com.thoughtworks.go.work.DefaultGoPublisher.reportCurrentStatus(DefaultGoPublisher.java:96)
	at com.thoughtworks.go.remote.work.BuildWork.reportCompletion(BuildWork.java:117)
	at com.thoughtworks.go.remote.work.BuildWork.doWork(BuildWork.java:94)
	at com.thoughtworks.go.agent.JobRunner.run(JobRunner.java:60)
	at com.thoughtworks.go.agent.AgentWebSocketClientController.process(AgentWebSocketClientController.java:130)
	at com.thoughtworks.go.agent.AgentWebSocketClientController$1.run(AgentWebSocketClientController.java:254)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by ibnc** [comment]
@jyotisingh 

> For legacy communication, the server waited for 10mins before marking the agent as lost contact. Although with web-sockets we have the capability to know instanteneouly when the connection dropped, may be it would make sense to retain the original behavior. What do you guys think?

That sounds reasonable to me.

-------

The last two issues I'm looking into. I suspect they're both related to the us using sendBytesByFuture.

**Comment by ibnc** [comment]
@jyotisingh 
Also, look at the timestamps for the acknowledgements. 

The agent reports that it's completing at "2017-04-28 07:19:35,137"

Then the agent gets the acknowledgement from the server at "2017-04-28 07:19:35,622"

So, we got an acknowledgement almost imminently. Here's the kicker, the thread the agent started didn't try to remove the callback until "2017-04-28 07:26:44,422" which is a little over 5 minutes. 

So, my question is why it took 5 minutes before we actually tried to remove the callback.

**Comment by jyotisingh** [comment]
@ibnc - yes, I noticed [that](https://github.com/gocd/gocd/issues/3438#issuecomment-298392915) as well. Not sure at this point.
One speculation is:
* Then the agent gets the acknowledgement from the server at "2017-04-28 07:19:35,622". That is logged [here](https://github.com/gocd/gocd/blob/9cec83fd6dda71ea085525354ca201ab7e4d3dec/agent/src/com/thoughtworks/go/agent/AgentWebSocketClientController.java#L249)
* [Executor](https://github.com/gocd/gocd/blob/9cec83fd6dda71ea085525354ca201ab7e4d3dec/agent/src/com/thoughtworks/go/agent/AgentWebSocketClientController.java#L239) is basically a thread pool of size 5. If all the 5 threads were processing other messages, then this one would have to wait before it reaches [here](https://github.com/gocd/gocd/blob/9cec83fd6dda71ea085525354ca201ab7e4d3dec/agent/src/com/thoughtworks/go/agent/AgentWebSocketClientController.java#L254). Could that be the reason why it took a while for the acknowledment message to be processed? I think we should add debug statements [here](https://github.com/gocd/gocd/blob/9cec83fd6dda71ea085525354ca201ab7e4d3dec/agent/src/com/thoughtworks/go/agent/AgentWebSocketClientController.java#L249), [here](https://github.com/gocd/gocd/blob/9cec83fd6dda71ea085525354ca201ab7e4d3dec/agent/src/com/thoughtworks/go/agent/AgentWebSocketClientController.java#L254) and [here]()

**Comment by ibnc** [comment]
@jyotisingh 

Yeah, that seems like the most probable explanation. I'll go ahead and add those debug statements, and open a PR.

**Comment by ibnc** [comment]
@jyotisingh Another possibility is that the timeout in the callback was exceeded before we could process the acknowledgement.

**Comment by ibnc** [comment]
@jyotisingh okay, I have a pretty good idea why the last error occurs ( the "RemoteEndpoint unavailable" one).

It happens when the websocket closes because of the idle timeout, and is in the process of reconnecting. While it's trying to reconnect, the session handler tries to send a message, but the websocket is not open. So, we need to ensure that the connection is actually open before we try to send. 

I'll work on a fix for this issue

**Comment by ibnc** [comment]
@jyotisingh After my most recent PR (https://github.com/gocd/gocd/pull/3487) gets merged, I believe this issue can be closed.

**Comment by arvindsv** [comment]
Closing. I believe we have or will be getting rid of websocket support for now, because it is not mature yet.
